Add /guilds endpoints

takos22 opened the pull request:
## Summary

I added `/guilds` endpoints and the models needed for it. I also updated the way tests are done and added tests that need a test database.

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] If endpoints were changed then they have been documented and tested.
    - [x] I have updated the docmentation to reflect the changes.
    - [x] I have updated the tests to support the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new endpoint or parameter).
- [ ] This PR is a breaking change (e.g. endpoint or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

takos22 committed:
🗃 Add Guild model

takos22 committed:
✨ Add guild blueprint

takos22 committed:
🗃 Add GuildConfig model

takos22 committed:
✨ Add /guilds endpoints

takos22 committed:
✨ Register guilds blueprint

takos22 committed:
✅ Update launch.py for easier testing

takos22 committed:
✅ Better testing

takos22 committed:
✅ Add tests for /guilds

takos22 committed:
🔊 Log every request

takos22 committed:
🗃 Update Guild.icon_url_as

takos22 committed:
✨ Add /guilds/<id>/icon endpoint

takos22 committed:
✅ Update test_guilds.py

takos22 committed:
👕 Update files for linter

takos22 committed:
Remove db tests in workflow

takos22 committed:
✅ Update tests

takos22 committed:
✅ Update tests

takos22 committed:
✨ Add /guilds/<id>/config endpoints
Update /guilds endpoints
Update GuildConfig model
Update tests

takos22 committed:
🐛 Fix bugs in /guilds

takos22 committed:
✨ Add VerifcationTypes enum in guild config

SylteA committed:
temp commit to work on other pc.

SylteA committed:
Merge remote-tracking branch 'origin/rewrite' into users

takos22 committed:
📚 Update docs
Add /guilds docs
Update /auth docs
Update docs template

takos22 committed:
Merge branch 'rewrite' into guilds

takos22 committed:
👕 Fix black linting

takos22 committed:
Merge branch 'guilds' of https://github.com/takos22/API into guilds

takos22 committed:
👕 I hate flake8

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,349 @@
+async def post_guild():
+    """Create a guild from the request body"""
+    data: dict = await request.json or {}
+    if not all(c in data for c in GUILD_COLUMNS):
```
Please switch to added validators.

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,349 @@
+    if guild is None:
+        # Guild already exists
+        return (
+            jsonify(
```
I'm working on standardizing the `jsonify` usage in our repository.

All usage of `jsonify` will be done like [this](https://github.com/Tech-With-Tim/API/blob/f83ce39e46fb058e33ff5c6a36671fa004363de9/utils/validators.py#L166-L169)

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,349 @@
+            409,
+        )
+
+    response = jsonify(
+        {
+            "id": str(guild.id),
+            "name": guild.name,
+            "owner_id": str(guild.owner_id),
+            "icon_hash": guild.icon_hash,
+        }
+    )
+    response.headers.add("Location", f"/guilds/{guild.id}")
+
+    return (
+        response,
+        201,
+    )
```
Instead of defining the response and adding headers, you can use [this ](https://pgjones.gitlab.io/quart/reference/response_values.html?highlight=return#combinations) syntax, it's much faster to develop like this and looks better.

takos22 replied:
I didnt know about this, thanks

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,349 @@
+
+    await guild.delete()
+
+    return jsonify(
```
Instead of returning data that's no longer in our database, you could really just return a `204 - No Content` response.

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,349 @@
+    return redirect(url)
+
+
+@bp.route("/<int:guild_id>/config", methods=["POST"])
```
(imo) GuildConfigs could be handled in a different view file (ex: `api/blueprints/guilds/views/configs.py`, overall cleaner this way.

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,156 @@
+        if not fields:
+            return self
+
+        allowed_fields = ("name", "owner_id", "icon_hash")
```
This should be handled in the view updating the guild. (Is handled by validator)

takos22 replied:
It is way simpler and cleaner like this (imo)

SylteA replied:
Validator will handle it anyways, no need to do it twice

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,156 @@
+    def icon_url_as(
+        self,
+        *,
+        format: Literal["jpeg", "jpg", "webp", "png", "gif"] = None,
```
please change `format` to something else (ex: `fmt`) as format is a builtin.

SylteA commented on launch.py:
```diff
@@ -76,23 +76,30 @@ def run_async(coro: Coroutine) -> Any:
     :param int retries:     Included to fix issue with docker starting API before DB is finished starting.
     :param float interval:  Interval of which to wait for next retry.
     """
+    DB_URI = db_uri or ENV["DB_URI"]
+    loop = loop or LOOP
```
Instead of making these optional then defining them like this, why not make it required and remove this?

takos22 committed:
Merge branch 'rewrite' of https://github.com/Tech-With-Tim/API into guilds

takos22 committed:
💚 Update lint and test workflow
Try to add postgres db to workflow

SylteA committed:
Add users endpoint.
add tests for validators
fix validators.py

takos22 committed:
Merge remote-tracking branch 'upstream/users' into guilds

takos22 committed:
✨ Update /guilds endpoints

takos22 committed:
✅ Update tests

takos22 committed:
📚 Update docs/README.md

takos22 committed:
💚 Update lint and test workflow for db

takos22 committed:
💚 Update lint and test workflow to postgres v13

SylteA commented on api/blueprints/guilds/views/configs.py:
```diff
@@ -0,0 +1,181 @@
+
+    try:
+        guild_config = await GuildConfig.create(guild_id, **data)
+    except ValueError as e:
```
This should never be raised as it's handled by validator

SylteA commented on api/blueprints/guilds/views/configs.py:
```diff
@@ -0,0 +1,181 @@
+import utils
+
+
+@bp.route("/<int:guild_id>/config", methods=["POST"])
```
remember to limit access (utils.app_only)

SylteA commented on api/blueprints/guilds/views/configs.py:
```diff
@@ -0,0 +1,181 @@
+    )
+
+
+@bp.route("/<int:guild_id>/config", methods=["GET"])
```
remember to limit access (utils.app_only)

SylteA commented on api/blueprints/guilds/views/configs.py:
```diff
@@ -0,0 +1,181 @@
+    )
+
+
+@bp.route("/<int:guild_id>/config", methods=["PATCH"])
```
remember to limit access (utils.app_only)

SylteA commented on api/blueprints/guilds/views/configs.py:
```diff
@@ -0,0 +1,181 @@
+    )
+
+
+@bp.route("/<int:guild_id>/config", methods=["DELETE"])
```
remember to limit access (utils.app_only)

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,148 @@
+request: utils.Request
+
+
+@bp.route("", methods=["POST"])
```
remember to limit access (utils.app_only)

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,148 @@
+    )
+
+
+@bp.route("/<int:guild_id>", methods=["PATCH"])
```
remember to limit access (utils.app_only)

SylteA commented on api/blueprints/guilds/views/guilds.py:
```diff
@@ -0,0 +1,148 @@
+    )
+
+
+@bp.route("/<int:guild_id>", methods=["DELETE"])
```
remember to limit access (utils.app_only)

SylteA committed:
Delete test_validators.py

To be included in a different branch

SylteA committed:
Delete users.py

To be included in a different branch

SylteA committed:
Delete __init__.py

To be included in a different branch

SylteA committed:
Delete __init__.py

To be included in a different branch

SylteA commented on tests/test_guilds.py:
```diff
@@ -0,0 +1,163 @@
+    assert response.status_code == status_code
+    if status_code == 201:
+        json = await response.json
+        assert int(json["id"]) == int(data["id"])
+        assert json["name"] == data["name"]
+        assert int(json["owner_id"]) == int(data["owner_id"])
+        assert json["icon_hash"] == data.get("icon_hash", None)
```
instead of asserting each keypair, assert the entire dict?

SylteA commented on tests/test_guilds.py:
```diff
@@ -0,0 +1,163 @@
+    )  # spamming random digits on keyboard
+    assert response.status_code == 404
+    assert response.content_type == "application/json"
+    json = await response.json
+    assert json["error"] == "Not found"
```
Instead of this, do 
```python
    assert (await response.json) == {...}
```

takos22 committed:
✨ Improve /guilds endpoints and tests

takos22 committed:
✨ Improve /guilds endpoints
Btw i hate flake8

takos22 committed:
💚 Update lint and test workflow
Add SECRET_KEY to env

takos22 committed:
✅ Update auth_app test fixture

takos22 committed:
✅ Update conftest.py

Remove unused import

takos22 committed:
✅ Update conftest.py
Formatting mhm

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,158 @@
+            return self
+
+        allowed_fields = ("name", "owner_id", "icon_hash")
+        fields = {
+            name: fields.get(name, getattr(self, name)) for name in allowed_fields
+        }
+
+        query = """
+        UPDATE guilds
+        SET (name, owner_id, icon_hash) = ($2, $3, $4)
+        WHERE id = $1
+        RETURNING *;
+        """
+        record = await self.pool.fetchrow(
+            query,
+            int(self.id),
+            fields["name"],
+            int(fields["owner_id"]),
+            fields["icon_hash"],
+        )
```
This will work, but is bad practice.

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,158 @@
+        if static_format not in VALID_STATIC_FORMATS:
+            raise ValueError(
+                "static_format must be one of {}".format(VALID_STATIC_FORMATS)
+            )
```
Please add a docstring and generally clean up this code.
Just adding a couple spaces would clean it up alot.

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,158 @@
+            )
+
+        if self.icon_hash is None:
+            return None
```
There are default icons even for guilds, we should return a default instead of None.

takos22 replied:
Are you sure about that?
This is when there is an icon:
![image](https://user-images.githubusercontent.com/59511469/106048673-79ab9f80-60e5-11eb-9052-951378f2ca45.png)
This is when there isn't an icon:
![image](https://user-images.githubusercontent.com/59511469/106048771-9647d780-60e5-11eb-96c4-755fe2570191.png)

SylteA commented on api/models/guild_config.py:
```diff
@@ -0,0 +1,172 @@
+    DISCORD_INTEGRATED_REACTION = "DISCORD_INTEGRATED_REACTION"
+
+
+class GuildConfig(Model):
```
Missing docstring.

SylteA commented on api/models/guild_config.py:
```diff
@@ -0,0 +1,172 @@
+
+        query = f"""
+        UPDATE guildconfigs
+        SET ({", ".join(allowed_fields)}) = ($2, $3, $4, $5, $6, $7, $8, $9, $10)
```
formatting a query like this is not very clean.
It's preferred if you did something like: 
```sql
UPDATE guildconfigs
SET
  xp_enabled = $2,
  xp_multiplier = $3,
  ...
WHERE guild_id = $1
RETURNING *;
```

This also counts for the other update method you have made.

SylteA commented on tests/test_guilds.py:
```diff
@@ -0,0 +1,141 @@
+@pytest.mark.db
+async def test_get_guild_404(app: QuartClient, db):
+    response = await app.get(
+        "/guilds/523456202403"
```
could of just written 0 :p

takos22 committed:
✨ Update /guilds enpoints and models
Add .fetch_or_404 for Guild and GuildConfig models
Add some docstrings
Clean up code

takos22 committed:
✨ Update /guilds endpoints
Update .fetch_or_404 mthods for easier use

takos22 committed:
🐛 Fix bugs in /guilds endpoints
Import configs in blueprint
Update configs endpoints

takos22 committed:
👕 Update formatting
I hate flake8

SylteA commented on README.md:
```diff
@@ -151,6 +151,6 @@ pipenv run pytest
 - [@SylteA](https://github.com/SylteA) - Most of the backend
 - [@Shubhaankar-sharma](https://github.com/Shubhaankar-sharma) - Docker deployment
-- [@takos22](https://github.com/takos22) - Markdown files
+- [@takos22](https://github.com/takos22) - /guilds endpoints and markdown files
```
I wasnt thinking we'd add info on who made what endpoints

SylteA commented on api/blueprints/guilds/views/configs.py:
```diff
@@ -0,0 +1,131 @@
+async def get_guild_config(guild_id: int):
+    """Get the Config for a guild"""
+
+    guild = await Guild.fetch_or_404(guild_id)
```
Why are you fetching the guild first?
There wont be a config if there is no guild.

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,183 @@
+    :param int id:          The guilds id.
+    :param str name:        The guilds name.
+    :param int owner_id:    The guilds owner id.
+    :param str icon_hash:   The guilds icon hash.
```
this is `Optional[str]`

SylteA commented on api/models/guild.py:
```diff
@@ -0,0 +1,183 @@
+
+        if guild := await cls.fetch(id):
+            return guild
+        else:
```
no need for else, no need for custom descriptions.

takos22 committed:
✨ Update /guilds endpoints
Fetch only config and not guild in /config endpoints
Update some docstrings
Update README.md

takos22 committed:
💚 Remove mixed-line-ending from pre-commit hook

SylteA reviewed (approved):
congrats 🥇
